README.md

```diff
@@ -1,13 +1,364 @@
+Mafia is an opinionated wrapper for Cabal that makes working on Haskell
+projects fun and easy.
 
 ![godfather](img/godfather.jpg)
+
+The central idea is that upon cloning a project, you should only have to
+run `mafia build` to get up and running for development. This will pull
+down any git submodules, create a cabal sandbox, install dependencies
+and then build everything, tests and benchmarks included.
+
+Cabal packages in git submodules are discovered automatically and made
+available as source dependencies. This is particularly useful if you
+have many internal dependencies that are not published to Hackage. It
+allows for development of a package and its dependencies all at once,
+without having to publish intermediate builds of libraries and without
+having to resort to a monorepo.
+
+All dependencies are cached globally in `$HOME/.ambiata/mafia/packages`
+using a nix-like hashing system so that for a given set of transitive
+dependencies, a package is only ever built once. Source dependencies are
+also cached, using their dependencies and source code as the hash. This
+is critical for Haskell development at scale. At Ambiata we have around
+150 Haskell packages which are being developed at a rapid pace, so it's
+crucial that we don't need to think about curating blessed sets of
+packages that we can cache as a single snapshot.
+
+
+How to use
+----------
+
+Mafia is designed to be run via a shell script which will bootstrap the
+`mafia` binary, but you can quite happily put the `mafia` binary on your
+`PATH` and use it like that as well.
+
+Place the
+[script](https://raw.githubusercontent.com/ambiata/mafia/topic/docs/script/mafia)
+in the same directory as your `.cabal` project file and then run
+`./mafia upgrade`. This will ensure that you have the latest script and
+also bake the git hash of the Mafia version in to the script.
+
+You can now run `./mafia build` for the first time. This will build the
+Mafia executable with the git hash specified in the script, and then
+build the current project. Having the hash baked in to the script means
+that anyone trying to build the project will be using exactly the same
+version of Mafia.
+
+To upgrade to the latest Mafia, you can `./mafia upgrade` again at any time.
+
+Mafia executables are stored in `$HOME/.ambiata/mafia/bin`.
+
+
+Commands
+--------
+
+Many of Mafia's commands are, more or less, direct wrappers of Cabal
+commands. The main difference is that they have a pre-step which takes
+care of installing any dependencies your project has, so you never have
+to think about it.
+
+### mafia build
+
+Builds a project, including all executables, test suites and benchmarks.
+
+Mafia uses whichever GHC it finds on the `PATH` to build a project. It
+keeps Cabal sandboxes in version specific directories, so that it is
+easy to [switch](https://gist.github.com/jystic/efded8238bc4c64affb7)
+between versions of GHC while developing a project.
```

Comment:
This doesn't work so well with the default setup for emacs haskell-mode, it gets confused about the different versions for say 7.8 and 7.10.